File: Python/Problems/Problems/problem1041.py
import heapq
import math
import sys

from utils import print1DArray, print2DArray
from typing import List, Set
from collections import Counter

from .problem import Problem


class Problem1041(Problem):
    def test(self) -> bool:
        instructions = "GL"

        expected = True

        result = self.isRobotBounded(instructions)

        return result == expected

    def isRobotBounded(self, instructions: str) -> bool:
        dirs = [0] * 4
        curr = 0
        for instruction in instructions:
            if instruction == 'G':
                dirs[curr] += 1
            elif instruction == 'R':
                curr = (curr + 1) % 4
            else:
                curr = (curr + 3) % 4

        return curr != 0 or dirs[0] == dirs[2] and dirs[1] == dirs[3]

File: C++/Problems/Problems/problem_1041.cpp
#include "problem_1041.h"

bool problem_1041::test() {
    std::string instructions = "GL";

    bool expected = true;

    auto result = isRobotBounded(instructions);

    return result == expected;
}

bool problem_1041::isRobotBounded(std::string instructions) {
    std::vector dirs(4, 0);
    int curr = 0;
    for (auto instruction : instructions)
    {
        if (instruction == 'G')
            dirs[curr]++;
        else if (instruction == 'R')
            curr = (curr + 1) % 4;
        else
            curr = (curr + 3) % 4;
    }


    return curr != 0 || dirs[0] == dirs[2] && dirs[1] == dirs[3];
}
